core/src/main/java/org/mule/runtime/core/util/collection/EventToMessageSequenceSplittingStrategy.java

```diff
@@ -39,35 +46,56 @@ public EventToMessageSequenceSplittingStrategy(ExpressionSplittingStrategy expre
+      }
+      if (payload instanceof Iterator<?>) {
+        return new IteratorMessageSequence(((Iterator<Object>) payload));
+      }
+      if (payload instanceof Collection) {
+        return new CollectionMessageSequence(copyCollection((Collection) payload));
+      } else if (payload instanceof CursorIteratorProvider) {
+        return new IteratorMessageSequence(((CursorIteratorProvider) payload).openCursor());
+      } else if (payload instanceof Iterable<?>) {
+        return new IteratorMessageSequence(((Iterable<Object>) payload).iterator());
+      } else if (payload instanceof Object[]) {
+        return new ArrayMessageSequence((Object[]) payload);
+      } else if (payload instanceof NodeList) {
+        return new NodeListMessageSequence((NodeList) payload);
+      } else if (payload instanceof Map<?, ?>) {
+        List<Object> list = new LinkedList<>();
+        Set<Map.Entry<?, ?>> set = ((Map) payload).entrySet();
+        for (Map.Entry<?, ?> entry : set) {
+          // TODO MULE-9502 Support "key" flowVar with MapSplitter in Mule 4
+          list.add(new ReadOnlyEntry<>(entry));
+        }
+        return new CollectionMessageSequence(list);
+      }
     }
-    if (payload instanceof Collection) {
-      return new CollectionMessageSequence(copyCollection((Collection) payload));
-    } else if (payload instanceof CursorIteratorProvider) {
-      return new IteratorMessageSequence(((CursorIteratorProvider) payload).openCursor());
-    } else if (payload instanceof Iterable<?>) {
-      return new IteratorMessageSequence(((Iterable<Object>) payload).iterator());
-    } else if (payload instanceof Object[]) {
-      return new ArrayMessageSequence((Object[]) payload);
-    } else if (payload instanceof NodeList) {
-      return new NodeListMessageSequence((NodeList) payload);
-    } else {
-      try {
-        // Let's try to split the payload using the expression manager since it may support based on the mimeType.
-        return new CollectionMessageSequence<>(expressionSplitterStrategy.split(event));
-      } catch (Exception e) {
-        throw new IllegalArgumentException(CoreMessages
-            .objectNotOfCorrectType(payload != null ? payload.getClass() : null,
-                                    new Class[] {Iterable.class, Iterator.class, MessageSequence.class, Collection.class})
-            .getMessage());
+    try {
+      Iterator<TypedValue<?>> valueIterator = expressionSplitterStrategy.split(event);
+      if (hasMelExpression(expressionSplitterStrategy.getExpression())) {
+        List<Object> iteratorCollection = new ArrayList<>();
+        valueIterator.forEachRemaining(iteratorCollection::add);
+        return new CollectionMessageSequence<>(iteratorCollection);
       }
+      return new IteratorMessageSequence(valueIterator);
+    } catch (ExpressionRuntimeException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
```

Comment:
This is how it was since 3.x. It's probably more descriptive than anything else that could happen. I just moved the code. For the case of ExpressionRuntimeException, we need to explicitly rethrow that exception so the error handling can put the right error type.